### PR TITLE
add a test for ingress SNI passthrough

### PR DESCRIPTION
**Pull request by linsun**
fixes: https://github.com/istio/istio.io/issues/7378

### CODE REVIEW

**Comment by linsun** [comment]
@frankbu PTAL.  The framework has been great - very easy to use!!

**Comment by linsun** [comment]
A bunch of lint issue (not discovered via lint-fast) I'll address

**Comment by frankbu** [comment]
@linsun Yes, somebody pointed out that `make lint-fast` doesn't seem to call the `shellcheck` linter any more, but haven't had a chance to figure out why. Apparently `make lint` will do it.

The lint errors in `snip_deploy_an_nginx_server_2` are a problem. The contents of the "here document" needs to be escaped so that the linter doesn't try to lint check the contents. The way to do it is to use `<<\EOF` instead of `<<EOF` on the first line, but the problem is that both the UI (display and copy button), and the snip generator don't handle that correctly. They only recognize `<<EOF`, so the UI and snip generator both need to be fixed to handle it. If you don't want to get into that right now, you can add a shellcheck disable comment (temporary fix), just before the command:
```
# shellcheck disable=SC2154
```

**Comment by linsun** [comment]
@frankbu thanks for the suggestion.  I went with the temp fix but getting this error on gen check now:

```
-# shellcheck disable=SC2154
 snip_deploy_an_nginx_server_2() {
 cat <<EOF > ./nginx.conf
 events {
ERROR: Some files need to be updated, please run 'make gen' and include any changed files in your PR
common/Makefile.common.mk:114: recipe for target 'check-clean-repo' failed
```

File: content/en/docs/tasks/traffic-management/ingress/ingress-sni-passthrough/index.md [hunk 1]
```diff
@@ -135,7 +136,7 @@ to hold the configuration of the NGINX server:
     printed correctly, i.e., `common name` is equal to `nginx.example.com`.
 
     {{< text bash >}}
-    $ kubectl exec -it $(kubectl get pod  -l run=my-nginx -o jsonpath={.items..metadata.name}) -c istio-proxy -- curl -v -k --resolve nginx.example.com:443:127.0.0.1 https://nginx.example.com
+    $ kubectl exec -it "$(kubectl get pod  -l run=my-nginx -o jsonpath={.items..metadata.name})" -c istio-proxy -- curl -v -k --resolve "nginx.example.com:443:127.0.0.1" "https://nginx.example.com"
```
Comment (frankbu) [inline content/en/docs/tasks/traffic-management/ingress/ingress-sni-passthrough/index.md:139]:
```suggestion
    $ kubectl exec -it "$(kubectl get pod  -l run=my-nginx -o jsonpath={.items..metadata.name})" -c istio-proxy -- curl -v -k --resolve nginx.example.com:443:127.0.0.1 https://nginx.example.com
```

File: content/en/docs/tasks/traffic-management/ingress/ingress-sni-passthrough/index.md [hunk 2]
```diff
@@ -225,7 +226,7 @@ to hold the configuration of the NGINX server:
     it is successfully verified (_SSL certificate verify ok_ is printed).
 
     {{< text bash >}}
-    $ curl -v --resolve nginx.example.com:$SECURE_INGRESS_PORT:$INGRESS_HOST --cacert example.com.crt https://nginx.example.com:$SECURE_INGRESS_PORT
+    $ curl -v --resolve "nginx.example.com:$SECURE_INGRESS_PORT:$INGRESS_HOST" --cacert "example.com.crt" "https://nginx.example.com:$SECURE_INGRESS_PORT"
```
Comment (frankbu) [inline content/en/docs/tasks/traffic-management/ingress/ingress-sni-passthrough/index.md:229]:
```suggestion
    $ curl -v --resolve "nginx.example.com:$SECURE_INGRESS_PORT:$INGRESS_HOST" --cacert example.com.crt "https://nginx.example.com:$SECURE_INGRESS_PORT"
```

File: tests/trafficmanagement/ingress/scripts/ingress_sni_passthrough.sh [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -e
+set -u
+set -o pipefail
+
+source "${REPO_ROOT}/content/en/docs/tasks/traffic-management/ingress/ingress_sni_passthrough/snips.sh"
```
Comment (frankbu) [inline tests/trafficmanagement/ingress/scripts/ingress_sni_passthrough.sh:22]:
```suggestion
source "${REPO_ROOT}/content/en/docs/tasks/traffic-management/ingress/ingress-sni-passthrough/snips.sh"
```

**Comment by frankbu** [comment]
@linsun you need to put the shellcheck comment in front of the command `cat <<EOF > ./nginx.conf` in the index.md file and then regen the snip file. Not so nice, but hopefully a temporary workaround until we fix the UI and generator to support escaped here docs.

File: tests/trafficmanagement/ingress/scripts/ingress_sni_passthrough.sh [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+sample_wait_for_deployment default nginx
+
+# validate NGINX server was deployed successfully 
+snip_deploy_an_nginx_server_5
+
+out=$(snip_deploy_an_nginx_server_5_out 2>&1)
+
+_verify_contains "$out" "common name: nginx.example.com (matched)"
+_verify_contains "$out" "server certificate expiration date OK"
+_verify_contains "$out" "issuer: O=example Inc.; CN=example.com"
```
Comment (frankbu) [inline tests/trafficmanagement/ingress/scripts/ingress_sni_passthrough.sh:50]:
```suggestion
_run_and_verify_contains snip_deploy_an_nginx_server_5 "common name: nginx.example.com (matched)"
```

File: tests/trafficmanagement/ingress/scripts/ingress_sni_passthrough.sh [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+snip_configure_an_ingress_gateway_1
+snip_configure_an_ingress_gateway_2
+snip_configure_an_ingress_gateway_3
+
+# validate the output
+out=$(snip_configure_an_ingress_gateway_3_out 2>&1)
+
+_verify_contains "$out" "subject: CN=nginx.example.com; O=some organization"
+_verify_contains "$out" "issuer: O=example Inc.; CN=example.com"
+_verify_contains "$out" "SSL certificate verify ok."
```
Comment (frankbu) [inline tests/trafficmanagement/ingress/scripts/ingress_sni_passthrough.sh:62]:
```suggestion
_run_and_verify_contains snip_configure_an_ingress_gateway_3 "SSL certificate verify ok."
```

File: tests/trafficmanagement/ingress/scripts/ingress_sni_passthrough.sh [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+out=$(snip_deploy_an_nginx_server_5_out 2>&1)
+
+_verify_contains "$out" "common name: nginx.example.com (matched)"
+_verify_contains "$out" "server certificate expiration date OK"
+_verify_contains "$out" "issuer: O=example Inc.; CN=example.com"
+
+# configure an ingress gateway
+snip_configure_an_ingress_gateway_1
+snip_configure_an_ingress_gateway_2
+snip_configure_an_ingress_gateway_3
```
Comment (frankbu) [inline tests/trafficmanagement/ingress/scripts/ingress_sni_passthrough.sh:55]:
```suggestion
```

File: tests/trafficmanagement/ingress/scripts/ingress_sni_passthrough.sh [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+snip_deploy_an_nginx_server_3
+
+snip_deploy_an_nginx_server_4
+
+# waiting for nginx deployment to start
+sample_wait_for_deployment default nginx
+
+# validate NGINX server was deployed successfully 
+snip_deploy_an_nginx_server_5
+
```
Comment (frankbu) [inline tests/trafficmanagement/ingress/scripts/ingress_sni_passthrough.sh:45]:
```suggestion
```

**Comment by frankbu** [comment]
@linsun I fixed the heredoc problem with the generator and UI in #7386 .
Now, instead of the `# shellcheck disable` workaround, you can fix it properly.

Change: `cat <<EOF > ./nginx.conf` to `cat <<\EOF > ./nginx.conf` in the index.md and then regen the snips. Lint should then be happy, and the UI copy button will still work.

**Comment by linsun** [comment]
@frankbu as an attempt to debug why test failed, I ran the test manually on my cluster.  Note: there is no longer _**common name: nginx.example.com (matched)**_ in the output.

```
$ kubectl exec -it $(kubectl get pod  -l run=my-nginx -o jsonpath={.items..metadata.name}) -c istio-proxy -- curl -v -k --resolve nginx.example.com:443:127.0.0.1 https://nginx.example.com


* Added nginx.example.com:443:127.0.0.1 to DNS cache
* Rebuilt URL to: https://nginx.example.com/
* Hostname nginx.example.com was found in DNS cache
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to nginx.example.com (127.0.0.1) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server accepted to use http/1.1
* Server certificate:
*  subject: CN=nginx.example.com; O=some organization
*  start date: May 27 14:18:47 2020 GMT
*  expire date: May 27 14:18:47 2021 GMT
*  issuer: O=example Inc.; CN=example.com
*  SSL certificate verify result: unable to get local issuer certificate (20), continuing anyway.
> GET / HTTP/1.1
> Host: nginx.example.com
> User-Agent: curl/7.58.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Server: nginx/1.17.10
< Date: Wed, 27 May 2020 15:36:19 GMT
< Content-Type: text/html
< Content-Length: 612
< Last-Modified: Tue, 14 Apr 2020 14:19:26 GMT
< Connection: keep-alive
< ETag: "5e95c66e-264"
< Accept-Ranges: bytes
< 
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
<style>
    body {
        width: 35em;
        margin: 0 auto;
        font-family: Tahoma, Verdana, Arial, sans-serif;
    }
</style>
</head>
<body>
<h1>Welcome to nginx!</h1>
<p>If you see this page, the nginx web server is successfully installed and
working. Further configuration is required.</p>

<p>For online documentation and support please refer to
<a href="http://nginx.org/">nginx.org</a>.<br/>
Commercial support is available at
<a href="http://nginx.com/">nginx.com</a>.</p>

<p><em>Thank you for using nginx.</em></p>
</body>
</html>
* Connection #0 to host nginx.example.com left intact
```

File: content/en/docs/tasks/traffic-management/ingress/ingress-sni-passthrough/index.md [hunk 8]
```diff
@@ -5,6 +5,7 @@ weight: 30
 keywords: [traffic-management,ingress,https]
 aliases:
   - /docs/examples/advanced-gateways/ingress-sni-passthrough/
+test: true
```
Comment (frankbu) [inline content/en/docs/tasks/traffic-management/ingress/ingress-sni-passthrough/index.md:8]:
```suggestion
test: yes
```

Comment (linsun) [reply content/en/docs/tasks/traffic-management/ingress/ingress-sni-passthrough/index.md]: updated

**Pull request merged**